### PR TITLE
API Implement cascade_deletes

### PR DESCRIPTION
**Pull request by tractorcow**
Framework component of https://github.com/silverstripe/silverstripe-versioned/issues/23

CMS fixes at https://github.com/silverstripe/silverstripe-cms/pull/1919

### CODE REVIEW

**Comment by tractorcow** [comment]
Looks like I need to take a double check of those CMS tests.

File: docs/en/02_Developer_Guides/00_Model/02_Relations.md [hunk 1]
```diff
@@ -406,6 +406,31 @@ the best way to think about it is that the object where the relationship will be
...
+    private static $cascade_deletes = [
+        'Child',
+    ];
+}
+class ChildObject extends DataObject {
+}
+```
+
+In this example, when the Parent object is deleted, the Child specified by the has_one relation will also
+be deleted.
```
Comment (flamerohr) [inline docs/en/02_Developer_Guides/00_Model/02_Relations.md]: Does this also apply to `many_many`, `belongs_many_many`, `has_many` and `many_many through` as well?
It seems to according to the docblock in `DataObject`
Perhaps add a note about those types of relationships here to be consistent.

File: src/ORM/DataObject.php [hunk 2]
```diff
@@ -1044,6 +1055,23 @@ protected function onAfterWrite()
     }
 
     /**
+     * Find all objects that will be cascade deleted if this object is deleted
+     *
+     * Notes:
+     *   - Objects will only be searched in the same stage as the given record.
```
Comment (flamerohr) [inline src/ORM/DataObject.php]: Add note:
> If this Object is versioned

File: src/ORM/DataObject.php [hunk 3]
```diff
@@ -3680,4 +3714,103 @@ public function setJoin(DataObject $object, $alias = null)
         }
         return $this;
     }
+
+    /**
+     * Find objects in the given relationships, merging them into the given list
+     *
+     * @param string $source Config property to extract relationships from
+     * @param bool $recursive True if recursive
+     * @param ArrayList $list Optional list to add items to
```
Comment (flamerohr) [inline src/ORM/DataObject.php]: is it more accurate to say this?
> list to add to for all items found

Maybe a note that this is used mainly to easily enable recursion?

File: src/ORM/DataObject.php [hunk 4]
```diff
@@ -3680,4 +3714,103 @@ public function setJoin(DataObject $object, $alias = null)
...
+
+        /** @var DataObject $item */
+        foreach ($items as $item) {
+            $this->mergeRelatedObject($list, $added, $item);
+        }
+        return $added;
+    }
+
+    /**
+     * Merge single object into a list
```
Comment (flamerohr) [inline src/ORM/DataObject.php]: Missing:
> if object already exists, it will not be added again

File: src/ORM/DataObject.php [hunk 5]
```diff
@@ -3680,4 +3714,103 @@ public function setJoin(DataObject $object, $alias = null)
...
+     *
+     * @param ArrayList $list Global list. Object will not be added if already added to this list.
+     * @param ArrayList $added Additional list to insert into
+     * @param DataObject $item Item to add
+     */
+    public function mergeRelatedObject($list, $added, $item)
```
Comment (flamerohr) [inline src/ORM/DataObject.php]: should this method be public? seemed to be quite a special case method

File: src/ORM/DataObject.php [hunk 6]
```diff
@@ -3680,4 +3714,103 @@ public function setJoin(DataObject $object, $alias = null)
...
+
+    /**
+     * Helper method to merge owned/owning items into a list.
+     * Items already present in the list will be skipped.
+     *
+     * @param ArrayList $list Items to merge into
+     * @param mixed $items List of new items to merge
+     * @return ArrayList List of all newly added items that did not already exist in $list
+     */
+    public function mergeRelatedObjects($list, $items)
```
Comment (flamerohr) [inline src/ORM/DataObject.php]: should this method be public?

File: src/ORM/DataObject.php [hunk 7]
```diff
@@ -3680,4 +3714,103 @@ public function setJoin(DataObject $object, $alias = null)
...
+
+    /**
+     * Find objects in the given relationships, merging them into the given list
+     *
+     * @param string $source Config property to extract relationships from
+     * @param bool $recursive True if recursive
+     * @param ArrayList $list Optional list to add items to
+     * @return ArrayList The list
+     */
+    public function findRelatedObjects($source, $recursive = true, $list = null)
```
Comment (flamerohr) [inline src/ORM/DataObject.php]: I'm not sure if this method needs to be public?

Comment (tractorcow) [inline src/ORM/DataObject.php]: It needs to be public because Versioned extension calls it.

File: src/ORM/DataObject.php (hunk 6 above)
Comment (tractorcow) [inline src/ORM/DataObject.php]: As above, Versioned extension needs to call it. It used to be protected and a member of Versioned, but it needs to be exposed now.

File: src/ORM/DataObject.php (hunk 5 above)
Comment (tractorcow) [inline src/ORM/DataObject.php]: Same as above.

Comment (tractorcow) [inline src/ORM/DataObject.php]: Ok, I think I can make just this one method protected, but not the others.

Comment (flamerohr) [inline src/ORM/DataObject.php]: Fair enough, I'm probably too critical about what's public, only half way through the Versioned pull request

**Comment by tractorcow** [comment]
Updated based on feedback.

File: src/ORM/DataObject.php (hunk 5 above)
Comment (flamerohr) [inline src/ORM/DataObject.php]: Just been through the other pull request, and yeah I agree that only `mergeRelatedObject()` could be made protected.

**Pull request merged**